test/utils/project_holder.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	boskosclient "sigs.k8s.io/boskos/client"
+	"sigs.k8s.io/boskos/common"
+)
+
+type ProjectHolder struct {
+	c *boskosclient.Client
+	// startCleanup channel signals when we need to stop refresh the resource and release the boskos project.
+	startCleanup chan bool
+	// finishCleanup channel signals when the clean up is finished.
+	finishCleanup chan bool
+}
+
+func NewProjectHolder() (*ProjectHolder, error) {
+	c, err := boskosclient.NewClient(os.Getenv("JOB_NAME"), "http://boskos", "", "")
+	if err != nil {
+		return nil, err
+	}
+	return &ProjectHolder{
+		c:             c,
+		startCleanup:  make(chan bool),
+		finishCleanup: make(chan bool),
+	}, nil
+}
+
+// Acquire tries to get a boskso project, and spawn a goroutine to refresh it.
+func (ph *ProjectHolder) Acquire(resourceType string) string {
+	// Try to get a Boskos project
+	klog.Info("Running in PROW")
+	klog.Info("Fetching a Boskos loaned project")
+
+	p := ph.getBoskosProject(resourceType)
+	project := p.Name
+	go ph.refresh(project)
+
+	return project
+}
+
+// Release stops the refresh goroutine, and release the boskso project.
+func (ph *ProjectHolder) Release(project string) {
+	ph.startCleanup <- true
+
+	// Wait until project cleanup is finished.
+	<-ph.finishCleanup
+}
+
+// Periodically refresh the resource to avoid the resource being cleaned
+// up accidentally.
+// Boskos Reaper component looks for resources that are owned but not
+// updated for a period of time, and resets stale resources to dirty state,
+// and Boskos Janitor component cleans up all dirty resources.
+func (ph *ProjectHolder) refresh(project string) {
+	for {
+		select {
+		case <-time.Tick(time.Minute * 5):
+			if err := ph.c.UpdateOne(project, "busy", nil); err != nil {
+				klog.Warningf("[Boskos] Update %s failed with %v", project, err)
+			}
+		case <-ph.startCleanup:
+			ph.c.ReleaseOne(project, "dirty")
+			ph.finishCleanup <- true
+			return
+		}
+	}
+}
+
+// getBoskosProject retries acquiring a boskos project until success or timeout.
```

Comment:
any methods that will terminate the process should be named XXXOrDie

getBoskosProjectOrDie

Reply:
Updated.